### DISCUSSION TITLE
maximum heap size reached when running testing with emqx

### DISCUSSION THREAD

**Discussion by sushantAtAws** [emqx/emqx#6704] [Q&A]
### Subject of the issue
Hi, 
I am running a long running test with emqx to tests its stability over days. I have a setup where EMQX is running in our VPC and listening on 8883. 10 publishers and 1 subscriber are connected to the broker and publishers are sending messages to the broker at 100 messages per second. Subscriber is getting these 1000 messages are asserting that. All at QoS 0.

Every thing works fine for ~3 days. EMQX memory uitilization is very consistent around 121MB for these 3 days. Suddenly the memory utilization peaks to around 164MB after the 3 days and it causes the OOM error. At this point, subscriber stops getting any messages.

OOM error in the logs: 
```
2022-01-07T13:28:41.354359+00:00 [error] Process:          <0.2016.0> on node 'emqx@127.0.0.1', Context:          maximum heap size reached, Max Heap Size:    20971520, Total Heap Size:  56519599, Kill:             true, Error Logger:     true, GC Info:          [{old_heap_block_size,22177879},{heap_block_size,22063419},{mbuf_size,12278323},{recent_size,3258},{stack_size,21},{old_heap_size,0},{heap_size,3581831},{bin_vheap_size,397945},{bin_vheap_block_size,514761},{bin_old_vheap_size,0},{bin_old_vheap_block_size,46422}].
```

I ran 2 separate tests and they both shows the exact same pattern after 3 days.
![image](https://user-images.githubusercontent.com/43186579/148660315-20a2ff53-41fa-4ae5-be4d-aeb2714be777.png)
![image](https://user-images.githubusercontent.com/43186579/148660369-018579a5-1ef5-41e9-9215-987bee4d76ae.png)

### Your environment

For EMQ X 4.3 or newer, please provide the log archive generated by [node_dump tool](https://github.com/emqx/emqx/blob/master/bin/node_dump) 
[node_dump_20220108_211405.tar.gz](https://github.com/emqx/emqx/files/7834038/node_dump_20220108_211405.tar.gz)

- EMQ X version (e.g. `emqx_ctl status`): 4.3.10
- Hardware configuration (e.g. `lscpu`):
```
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz
Stepping:                        4
CPU MHz:                         2500.004
BogoMIPS:                        5000.00
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        1 MiB
L3 cache:                        33 MiB
```
- OS (e.g. `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
- Kernel (e.g. `uname -a`):
```
Linux ip-10-0-161-254 5.11.0-1022-aws #23~20.04.1-Ubuntu SMP Mon Nov 15 14:03:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
- Erlang/OTP version (in case you build emqx from source code): Erlang/OTP 22 [erts-10.6.4]
- Others:

### Steps to reproduce
Tell us how to reproduce this issue. 

### Expected behaviour
The process should not go OOM.

### Actual behaviour
The process goes OOM.

Missing log file can delay the handling of the issue.

**Comment by sushantAtAws** [comment]
Update: 
The broker is still running. I kicked off another subscriber to connect with the same instance of the broker and the subscriber started getting messages. 

I still would like to understand the bahavior of the broker and why it went OOM and recovered. This is not even a QoS 1 test case, it is QoS 0.

**Comment by zmstone** [comment]
Hi @sushantAtAws 
I am converting this to a discussion.

**Comment by zmstone** [comment]
Hi @sushantAtAws 
The OOM threshold is controlled by this config
`zone.$name.force_shutdown_policy=10000|64MB`
Maybe you can increase the 64MB part.

The heap size is not only affected by the queued messages,
but also the runtime garbage accumulated during the process' life-time.
The exact behaviour of garbage collector is hard to predict,
e.g. when the subscriber TCP buffer is not keeping up, the push-back may lead
to messages queued in the process mailbox (which is a VM runtime queue, 
but not the so called `mqueue` in EMQ X which is aware of QoS)
The easy way forward is maybe to increase max heap size.

**Reply by Gypsying** [reply]
> Hi @sushantAtAws The OOM threshold is controlled by this config `zone.$name.force_shutdown_policy=10000|64MB` Maybe you can increase the 64MB part.
> 
> The heap size is not only affected by the queued messages, but also the runtime garbage accumulated during the process' life-time. The exact behaviour of garbage collector is hard to predict, e.g. when the subscriber TCP buffer is not keeping up, the push-back may lead to messages queued in the process mailbox (which is a VM runtime queue, but not the so called `mqueue` in EMQ X which is aware of QoS) The easy way forward is maybe to increase max heap size.


Hi @zmstone 
The bahavior looks like [issue](https://github.com/emqx/emqx/issues/8765) which i encountered...

**Comment by kadarpik** [comment]
I think we need possibility to change heap size here, config file setting does not change anything, I see the same max heap sizes from logs.
